src/darwin/Framework/CHIP/MTRDeviceConnectivityMonitor.mm

```diff
@@ -0,0 +1,247 @@
+            MTR_LOG_ERROR("%@ cannot set dispatch queue on resolve", self);
+            DNSServiceRefDeallocate(sSharedResolverConnection);
+            sSharedResolverConnection = NULL;
+            sSharedResolverQueue = nil;
+            return NULL;
+        }
+    }
+
+    return sSharedResolverConnection;
+}
+
+- (void)_callHandler
+{
+    os_unfair_lock_assert_owner(&sConnectivityMonitorLock);
+    MTRDeviceConnectivityMonitorHandler handlerToCall = self->_monitorHandler;
+    if (handlerToCall) {
+        dispatch_async(self->_handlerQueue, ^{ handlerToCall(); });
+    }
+}
+
+- (void)handleResolvedHostname:(const char *)hostName port:(uint16_t)port error:(DNSServiceErrorType)error
+{
+    std::lock_guard lock(sConnectivityMonitorLock);
+
+    // dns_sd.h: must check and call deallocate if error is kDNSServiceErr_ServiceNotRunning
+    if (error == kDNSServiceErr_ServiceNotRunning) {
+        MTR_LOG_ERROR("%@ disconnected from dns-sd subsystem", self);
+        [self _stopMonitoring];
+        return;
+    }
+
+    // Create a nw_connection to monitor connectivity if the host name is not being monitored yet
+    NSString * hostNameString = [NSString stringWithUTF8String:hostName];
+    if (!_connections[hostNameString]) {
+        char portString[6];
+        snprintf(portString, sizeof(portString), "%d", ntohs(port));
+        nw_endpoint_t endpoint = nw_endpoint_create_host(hostName, portString);
+        if (!endpoint) {
+            MTR_LOG_ERROR("%@ failed to create endpoint for %s:%s", self, hostName, portString);
+            return;
+        }
+        nw_parameters_t params = nw_parameters_create_secure_udp(NW_PARAMETERS_DISABLE_PROTOCOL, NW_PARAMETERS_DEFAULT_CONFIGURATION);
+        if (!params) {
+            MTR_LOG_ERROR("%@ failed to create udp parameters", self);
+            return;
+        }
+        nw_connection_t connection = nw_connection_create(endpoint, params);
+        if (!connection) {
+            MTR_LOG_ERROR("%@ failed to create connection for %s:%s", self, hostName, portString);
+            return;
+        }
+        nw_connection_set_queue(connection, sSharedResolverQueue);
+        nw_connection_set_path_changed_handler(connection, ^(nw_path_t _Nonnull path) {
+            nw_path_status_t status = nw_path_get_status(path);
+            if (status == nw_path_status_satisfied) {
+                MTR_LOG_INFO("%@ path is satisfied", self);
+                std::lock_guard lock(sConnectivityMonitorLock);
+                [self _callHandler];
+            }
+        });
+        nw_connection_set_viability_changed_handler(connection, ^(bool viable) {
+            if (viable) {
+                std::lock_guard lock(sConnectivityMonitorLock);
+                MTR_LOG_INFO("%@ connectivity now viable", self);
+                [self _callHandler];
+            }
+        });
+        nw_connection_start(connection);
+    }
+}
+
+static void _resolveReplyCallback(
```

Comment:
```suggestion
static void ResolveCallback(
```
Nit: `_resolveReplyCallback` looks confusingly like an ivar where it's used.

Reply:
Alternatively consider using a lambda without captures.

Reply:
Done